### PR TITLE
import route

### PR DESCRIPTION
**Pull request by svub**
+ RecoveryWordsInput component 
+ AutoComplete lib 
+ AnimationUtils lib 
+ Privacy Agent component

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
       "integrity": "sha1-63d99gEXI6OxTopywIBcjoZ0a9I=",
       "dev": true,
       "requires": {
-        "mime-types": "2.1.18",
+        "mime-types": "~2.1.18",
```
Comment (sisou) [inline package-lock.json]: If you do not change the package.json, there is no need to update this file in the PR. Please remove changes to this file.

File: src/common.js [hunk 2]
```diff
@@ -1,3 +1,6 @@
+/* global TRANSLATIONS */
+I18n.initialize(TRANSLATIONS, 'en');
```
Comment (sisou) [inline src/common.js]: This is fixed in Max' PR, please remove here.

File: src/components/PrivacyAgent.css [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    font-weight: 500;
+    font-size: 24px;
+}
+
+.privacy-agent p {
+    max-width: 400px;
+    font-size: 15px;
+}
+
+.privacy-agent svg{
```
Comment (sisou) [inline src/components/PrivacyAgent.css]: Missing space before `{`.

File: src/components/PrivacyAgent.css [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    opacity: 0.8;
+    display: block;
+}
+
+.privacy-agent button{
```
Comment (sisou) [inline src/components/PrivacyAgent.css]: Missing space before `{`.

File: src/components/PrivacyAgent.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+class PrivacyAgent { // eslint-disable-line no-unused-vars
+    /**
+     * @param {string} nexthash - the hash to route to when the user clicked "ok"
+     * @param {HTMLElement} [el]
+     */
+    constructor(nexthash = 'continue', el) {
```
Comment (sisou) [inline src/components/PrivacyAgent.js]: Better to use `Nimiq.Observable` and to `fire` an event when the user clicks the "accept" button, like other components. The component should not be required to know about the environment that it is used in.

File: src/components/PrivacyAgent.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        this.$el = el || this._createElement(nexthash);
```
Comment (sisou) [inline src/components/PrivacyAgent.js]: Forward the el to the _createElement function to use as optional uninitialized root element, just like in `RecoverWordsInput` component.

File: src/components/RecoveryWordsInput.css [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    background: url('data:image/svg+xml;base64,PHN2ZyBmaWxsPSIjZmZmIiBoZWlnaHQ9IjI0IiB2aWV3Qm94PSIwIDAgMjQgMjQiIHdpZHRoPSIyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cGF0aCBkPSJNMCAwaDI0djI0SDB6IiBmaWxsPSJub25lIi8+PHBhdGggZD0iTTE4IDhoLTFWNmMwLTIuNzYtMi4yNC01LTUtNVM3IDMuMjQgNyA2djJINmMtMS4xIDAtMiAuOS0yIDJ2MTBjMCAxLjEuOSAyIDIgMmgxMmMxLjEgMCAyLS45IDItMlYxMGMwLTEuMS0uOS0yLTItMnptLTYgOWMtMS4xIDAtMi0uOS0yLTJzLjktMiAyLTIgMiAuOSAyIDItLjkgMi0yIDJ6bTMuMS05SDguOVY2YzAtMS43MSAxLjM5LTMuMSAzLjEtMy4xIDEuNzEgMCAzLjEgMS4zOSAzLjEgMy4xdjJ6Ii8+PC9zdmc+');
+    background-size: 14px;
+    background-position: 9px center;
+    background-repeat: no-repeat;
+    background-color: black;
+    padding-left: 4rem;
+    padding-right: 2rem;
+}
+
+.recovery-words-input .recovery-words-input-field {
```
Comment (sisou) [inline src/components/RecoveryWordsInput.css]: Are we still using this animation?

File: src/components/RecoveryWordsInputField.css [hunk 8]
```diff
@@ -0,0 +1,7 @@
+.recovery-word-field{
+    width: 100%;
+    box-sizing: border-box;
+  }
+
+ .recovery-word-field.shake input {
+      border-bottom: 1px solid var(--error-color) !important;
```
Comment (sisou) [inline src/components/RecoveryWordsInputField.css]: Missing closing bracket.

File: src/components/RecoveryWordsInputField.css [hunk 9]
```diff
@@ -0,0 +1,7 @@
+.recovery-word-field{
```
Comment (sisou) [inline src/components/RecoveryWordsInputField.css]: Missing space before `{`.

File: src/components/RecoveryWordsInputField.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        const setPlaceholder = () => {
+            input.placeholder = `${I18n.translatePhrase('recovery-words-input-field-placeholder')}${this._index + 1}`;
+        };
+        I18n.observer.on(I18n.Events.LANGUAGE_CHANGED, setPlaceholder);
+        setPlaceholder();
+
+        input.addEventListener('keydown', this._onKeydown.bind(this));
+        input.addEventListener('blur', this._onBlur.bind(this));
+        /** @type {HTMLInputElement} */
+        this.$input = input;
```
Comment (sisou) [inline src/components/RecoveryWordsInputField.js]: You are setting `this.$input` here, but are also returning it and setting it as `this.$.input`. Same for the placeholder, but not for `element`. Please decide for one approach.

File: src/components/RecoveryWordsInputField.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                response(list);
+            },
+            onSelect: this._focusNext.bind(this),
+            minChars: 3,
+            delay: 0,
+        });
+    }
+
+    focus() {
+        requestAnimationFrame(() => this.$.input.focus());
```
Comment (sisou) [inline src/components/RecoveryWordsInputField.js]: `requestAnimationFrame` executes instantly if the browser has capacity. What we want here is `setTimeout` without a timeout parameter (or `0`).

File: src/components/RecoveryWordsInputField.js [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+        // don't hide focused input fields
+        if (document.activeElement === this.$input) return;
+
+        this.$.element.classList.add('has-placeholder');
+        this.$.element.replaceChild(this.$.placeholder, this.$.input);
+    }
+
+    _showInput() {
+        // if (this._revealedWord === target || !target.classList.contains('has-placeholder')) return;
```
Comment (sisou) [inline src/components/RecoveryWordsInputField.js]: Remove unused comment.

File: src/lib/AutoComplete.js [hunk 13]
```diff
@@ -0,0 +1,243 @@
+/*
```
Comment (sisou) [inline src/lib/AutoComplete.js]: Note to other reviewers: This lib was not written by us, but taken from the below source. It was only adapted to satisfy our typescript and eslint rules.

File: src/lib/MnemonicPhrase.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
+class MnemonicPhrase {
```
Comment (sisou) [inline src/lib/MnemonicPhrase.js]: Note to other reviewers: This file was already approved in PR https://github.com/nimiq/keyguard-next/pull/17.

File: src/request/import/words/ImportWordsApi.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    async _handlePassphraseInput(passphrase) {
+        if (!this._privateKey) {
+            throw new Error('Private key not set!');
+        }
+
+        document.body.classList.add('loading');
+
+        const keyPair = Nimiq.KeyPair.derive(this._privateKey);
+        // const key = await Key.loadPlain(keyPair.serialize(), EncryptionType.HIGH);
```
Comment (sisou) [inline src/request/import/words/ImportWordsApi.js]: Unneeded comment?

File: src/request/import/words/ImportWordsApi.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    async onRequest() {
+        // show UI
+        window.location.hash = ImportWordsApi.Pages.PRIVACY_AGENT;
+    }
+
+    _makeView() {
+        this.$rootElement = /** @type {HTMLElement} */ (document.getElementById('app'));
+        this.$privacyAgent = /** @type {HTMLElement} */ (this.$rootElement.querySelector('#privacy'));
+        this.$enterWords = /** @type {HTMLElement} */ (this.$rootElement.querySelector('#words'));
+        this.$enterPassphrase = /** @type {HTMLElement} */ (this.$rootElement.querySelector('#passphrase'));
```
Comment (Bettelstab) [inline src/request/import/words/ImportWordsApi.js]: assign locally

File: src/translations/index.js [hunk 17]
```diff
@@ -16,7 +23,13 @@ const TRANSLATIONS = {
         'passphrase-placeholder': 'Passphrase eingeben',
         'sign-tx-wrong-passphrase': 'Falsche Passphrase, bitte versuche es nochmal',
         'sign-tx-button-enter-pin': 'PIN eingeben',
+        'privacy-agent-headline': 'Werden sie beobachtet?',
+        'privacy-agent-info': 'Jetzt ist eine gute Zeit um sich einmal umzuschauen. Gibt es Fenster in der Nähe? '
+                            + ' Versteckte Kameras? Jemand der über die Schulter schaut?',
+        'privacy-agent-warning': 'Falls irgendjemand ihre Wiederherstellungswörter herausfindet '
+                               + 'kann er ihre gesamten NIM stehlen!',
+        'privacy-agent-ok': 'OK, all good',
+        'recovery-words-input-label': 'Wiederherstellungsworte',
```
Comment (Bettelstab) [inline src/translations/index.js]: Wörter ;)

File: src/translations/index.js [hunk 18]
```diff
@@ -16,7 +23,13 @@ const TRANSLATIONS = {
         'passphrase-placeholder': 'Passphrase eingeben',
         'sign-tx-wrong-passphrase': 'Falsche Passphrase, bitte versuche es nochmal',
         'sign-tx-button-enter-pin': 'PIN eingeben',
+        'privacy-agent-headline': 'Werden sie beobachtet?',
+        'privacy-agent-info': 'Jetzt ist eine gute Zeit um sich einmal umzuschauen. Gibt es Fenster in der Nähe? '
+                            + ' Versteckte Kameras? Jemand der über die Schulter schaut?',
+        'privacy-agent-warning': 'Falls irgendjemand ihre Wiederherstellungswörter herausfindet '
+                               + 'kann er ihre gesamten NIM stehlen!',
+        'privacy-agent-ok': 'OK, all good',
```
Comment (Bettelstab) [inline src/translations/index.js]: englisch?!

File: src/request/import/words/index.html [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    <script src="../../../translations/index.js"></script>
+    <script src="../../../common.js"></script>
+    <script src="../../../components/Identicon.js"></script>
+    <script src="../../../components/PrivacyAgent.js"></script>
+    <script src="../../../components/RecoveryWordsInput.js"></script>
+    <script src="../../../components/RecoveryWordsInputField.js"></script>
+    <script src="../../../components/PassphraseInput.js"></script>
+    <script src="../../PopupApi.js"></script>
+    <script src="ImportWordsApi.js"></script>
+    <script src="index.js"></script>
```
Comment (Bettelstab) [inline src/request/import/words/index.html]: add defer

File: src/components/RecoveryWordsInputField.js (hunk 11 above)
Comment (svub) [reply src/components/RecoveryWordsInputField.js]: small numbers cause trouble in Firefox (https://stackoverflow.com/questions/20747591/why-doesnt-this-settimeout-based-code-work-in-firefox-with-a-small-timeout-wor) - I'll put a small number there.

File: demos/RecoveryWordsInput.html [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+    <link href="../src/components/RecoveryWordsInput.css" rel="stylesheet">
+    <script src="http://cdn.nimiq.com/web-offline.js"></script>
+    <script src="../src/lib/MnemonicPhrase.js"></script>
+    <script src="../src/lib/I18n.js"></script>
+    <script src="../src/translations/index.js"></script>
+    <script src="../src/lib/AutoComplete.js"></script>
+    <script src="../src/lib/AnimationUtils.js"></script>
+    <script src="../src/components/RecoveryWordsInput.js"></script>
+    <script src="../src/components/RecoveryWordsInputField.js"></script>
+    <script src="../src/components/LanguagePicker.js"></script>
```
[the remark this answers is not in the document]
Comment (svub) [reply demos/RecoveryWordsInput.html]: for the demo page?

[the remark this answers is not in the document]
Comment (Bettelstab) [reply demos/RecoveryWordsInput.html]: shouldn't the demo page work as intended? :D

File: demos/RecoveryWordsInput.html [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        .button-row {
+            margin-top: 1em;
+        }
+
+        #privateKey {
+            margin-top: 2em;
+        }
+    </style>
```
[the remark this answers is not in the document]
Comment (svub) [reply demos/RecoveryWordsInput.html]: files mergerd

File: src/components/RecoveryWordsInput.css (hunk 7 above)
Comment (svub) [reply src/components/RecoveryWordsInput.css]: yep, the fields slide in "on show". I'd say it's optional

File: src/components/RecoveryWordsInputField.js (hunk 10 above)
Comment (svub) [reply src/components/RecoveryWordsInputField.js]: :+1:

File: src/lib/I18n.js [hunk 22]
```diff
@@ -120,10 +118,23 @@ class I18n { // eslint-disable-line no-unused-vars
...
+            // eslint-disable-next-line no-console
             console.warn(`Language ${language} not supported, using ${languageToUse} instead.`);
         }
 
-        this._language = languageToUse;
+        if (this._language !== languageToUse) {
+            /** @type {string} */
+            this._language = languageToUse;
+
+            if (({ interactive: 1, complete: 1 })[document.readyState]) {
```
[the remark this answers is not in the document]
Comment (svub) [reply src/lib/I18n.js]: will update in i18n branch, not here, ok? :)

File: src/components/PrivacyAgent.js [hunk 23]
```diff
@@ -39,20 +43,28 @@ class PrivacyAgent { // eslint-disable-line no-unused-vars
...
             <div class="grow"></div>
-            <a href="#continue"><button data-i18n="privacy-agent-ok">OK, all good</button></a>
+            <button data-i18n="privacy-agent-ok">OK, all good</button>
         `;
-        this.$ok = el.querySelector('a');
-        if (!this.$ok) {
+        /* eslint-enable max-len */
+
+        const $ok = element.querySelector('button');
+        if (!$ok) {
```
Comment (Bettelstab) [inline src/components/PrivacyAgent.js]: Just cast to HTMLButtonElement. We know it's there.

Comment (sisou) [reply src/components/PrivacyAgent.js]: @svub like this:
```
/** @type {HTMLButtonElement} /*
const $ok = (element.querySelector('button'));
```

File: src/request/import-words/ImportWordsApi.js [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+
+        const recoveryWordsInput = new RecoveryWordsInput();
+        // for debugging
+        // window.input = recoveryWordsInput;
+        recoveryWordsInput.on(RecoveryWordsInput.Events.COMPLETE, this._onRecoveryWordsEntered.bind(this));
+        $enterWords.appendChild(recoveryWordsInput.getElement());
+
+        const passphraseInput = new PassphraseInput(true);
+        passphraseInput.on(PassphraseInput.Events.PASSPHRASE_ENTERED, this._handlePassphraseInput.bind(this));
+        $enterPassphrase.appendChild(passphraseInput.getElement());
```
Comment (danimoh) [inline src/request/import-words/ImportWordsApi.js]: Does the user also have to repeat the passphrase?

File: src/components/RecoveryWordsInput.js [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+        label.innerHTML = '<div data-i18n="recovery-words-input-label"></div>';
+        el.appendChild(label);
+
+        const form = document.createElement('form');
+        form.setAttribute('autocomplete', 'off');
+        el.appendChild(form);
+
+        for (let index = 0; index < 24; index++) {
+            const field = new RecoveryWordsInputField(index);
+            field.element.dataset.index = index.toString();
```
[the remark this answers is not in the document]
Comment (svub) [reply src/components/RecoveryWordsInput.js]: just had a quick talk with Daniel: let's use "get element" everywhere!

[the remark this answers is not in the document]
Comment (svub) [reply src/components/RecoveryWordsInput.js]: @sisou @Bettelstab, agreed? :)

File: src/components/RecoveryWordsInput.js [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+        const check = this.$fields.find(field => !field.complete);
+        if (typeof check !== 'undefined') return;
+
+        const words = this.$fields.map(field => field.value).join(' ');
+        try {
+            this.fire(RecoveryWordsInput.Events.COMPLETE, words);
+        } catch (e) {
+            // assume wrong words but printing out for debugging
+            console.error(e); // eslint-disable-line no-console
+            this._animateError();
```
[the remark this answers is not in the document]
Comment (svub) [reply src/components/RecoveryWordsInput.js]: true! included into the component after reviewing the situation

File: src/components/RecoveryWordsInput.js [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+            const field = new RecoveryWordsInputField(index);
+            field.element.dataset.index = index.toString();
+            field.on(RecoveryWordsInputField.Events.VALID, this._onFieldComplete.bind(this));
+            field.on(RecoveryWordsInputField.Events.FOCUS_NEXT, this._setFocusToNextInput.bind(this));
+
+            this.$fields.push(field);
+            form.appendChild(field.element);
+        }
+
+        setTimeout(() => this.$fields[0].focus(), 100);
```
[the remark this answers is not in the document]
Comment (svub) [reply src/components/RecoveryWordsInput.js]: Improved code by using `this.focus()`---if used somewhere else, `input.focus()` needs to be called when being made visible. Added `input.focus()` in `ImportWordsApi` when switching to input view. works.

File: src/lib/AutoComplete.js [hunk 28]
```diff
@@ -0,0 +1,241 @@
...
+            delay: 150,
+            offsetLeft: 0,
+            offsetTop: 1,
+            cache: 1,
+            menuClass: '',
+            renderItem: /** @param {string} item @param {string} search */ (item, search) => {
+                // escape special characters
+                search = search.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
+                const re = new RegExp(`(${search.split(' ').join('|')})`, 'gi');
+                return `<div class="autocomplete-suggestion" data-val="${item}">${item.replace(re, '<b>$1</b>')}</div>`;
```
[the remark this answers is not in the document]
Comment (svub) [reply src/lib/AutoComplete.js]: true, using `createElement` and `textContent`.

File: src/request/import-words/ImportWordsApi.js (hunk 24 above)
Comment (svub) [reply src/request/import-words/ImportWordsApi.js]: ooookaaaay :P

**Pull request merged**